### PR TITLE
CLOUDSTACK-9560 Root volume of deleted VM left unremoved

### PR DESCRIPTION
**Pull request by yvsubhash**

### CODE REVIEW

**Comment by syed** [comment]
LGTM :+1:

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

File: server/src/com/cloud/storage/StorageManagerImpl.java [hunk 1]
```diff
@@ -2199,15 +2199,20 @@ public void cleanupDownloadUrls(){
...
+                s_logger.debug("Removing download url " + volumeOnImageStore.getExtractUrl() + " for volume id " + volumeId);
 
                 // Remove it from image store
                 ImageStoreEntity secStore = (ImageStoreEntity) _dataStoreMgr.getDataStore(volumeOnImageStore.getDataStoreId(), DataStoreRole.Image);
                 secStore.deleteExtractUrl(volumeOnImageStore.getInstallPath(), volumeOnImageStore.getExtractUrl(), Upload.Type.VOLUME);
 
                 // Now expunge it from DB since this entry was created only for download purpose
                 _volumeStoreDao.expunge(volumeOnImageStore.getId());
+                Volume volume = _volumeDao.findById(volumeId);
+                if (volume.getState() == Volume.State.Expunged)
```
Comment (koushik-das) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: Null check on volume?

**Comment by yvsubhash** [comment]
@koushik-das Added the null check

**Comment by koushik-das** [comment]
Code LGTM

File: server/src/com/cloud/storage/StorageManagerImpl.java [hunk 2]
```diff
@@ -2199,15 +2199,20 @@ public void cleanupDownloadUrls(){
...
                 ImageStoreEntity secStore = (ImageStoreEntity) _dataStoreMgr.getDataStore(volumeOnImageStore.getDataStoreId(), DataStoreRole.Image);
                 secStore.deleteExtractUrl(volumeOnImageStore.getInstallPath(), volumeOnImageStore.getExtractUrl(), Upload.Type.VOLUME);
 
                 // Now expunge it from DB since this entry was created only for download purpose
                 _volumeStoreDao.expunge(volumeOnImageStore.getId());
+                Volume volume = _volumeDao.findById(volumeId);
+                if (volume != null && volume.getState() == Volume.State.Expunged)
+                {
+                    _volumeDao.remove(volumeId);
+                }
```
Comment (ustcweizhou) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: There might be multiple volume snapshot on image store.
We need to make sure that all volume snapshot have been removed from image store.

```
+                VolumeDataStoreVO volumesOnImageStore = _volumeStoreDao.findByVolume(volumeId);
+                if (volumesOnImageStore == null) { // volume is expunged from primary, as well as on secondary
+                    VolumeVO vol = _volumeDao.findById(volumeId);
+                    if (vol != null && (vol.getState() == Volume.State.Expunged)) {
+                        _volumeDao.remove(volumeId);
+                    }
+                }
```

Comment (rhtyd) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: @yvsubhash can you reply to @ustcweizhou 's comment above ^^ ?

Comment (jburwell) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: @yvsubhash have you had a chance to review @ustcweizhou's feedback?

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@yvsubhash can you reply to raised review comments?

File: server/src/com/cloud/storage/StorageManagerImpl.java (hunk 2 above)
Comment (rhtyd) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: @yvsubhash ping

**Comment by yvsubhash** [comment]
@ustcweizhou  Volume snapshots would be left over even in case of normal vm destroy and that is expected. They can be used if there is a need to restore the volume at a later point in time.  The snapshots are visible under storage->snapshots. So i dont see a need to clean up those snapshots

**Comment by yvsubhash** [comment]
@rhtyd  Can you please merge this

File: server/src/com/cloud/storage/StorageManagerImpl.java (hunk 2 above)
Comment (rhtyd) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: @yvsubhash can you have a look at @ustcweizhou 's comment.

**Comment by ustcweizhou** [comment]
@yvsubhash I did not mean we need clean up the snapshots. 
in my understanding, if the volume has snapshots left, it is Expunged, but not marked as Removed
your change just delete one snapshot and then mark the volume as Removed, however, there might be other snapshots left (in this case, we should not mark the volume as Removed)

**Comment by yvsubhash** [comment]
@ustcweizhou  In case of volume snapsnots, volume gets set to removed irrespective of the presence of snapshots. I have verified the same  just now and I am only removing volume not any of the snapshots

**Comment by ustcweizhou** [comment]
@yvsubhash so, the question is , in what condition the volume is Expunged but not removed ? if it does not happen, then the change from line 2211 to 2215 is not needed.
however, to be honest, I often see a volume is Expunged but not Removed.

**Comment by yvsubhash** [comment]
@ustcweizhou  As I understand, Volume goes to Expunged state when it is removed from primary and gets removed when it is removed from secondary storage. line 2211 to 2215 is to handle specific case where volume getting removed from primary But not getting removed from secondary because of the download link. So cleaning up the same after the downlink is expired

**Comment by ustcweizhou** [comment]
@yvsubhash yes, you are right, it is 'download links', not 'volume snapshots'. I made changes on this long time ago and almost forgot how it works.

so, my question is, should we 'Remove' the volume after a download link is expired, or all download links are expired ?

**Comment by yvsubhash** [comment]
@ustcweizhou  thats good point. It should be removed after all download links expire. I will address this case

**Comment by yvsubhash** [comment]
@ustcweizhou The changed suggested by you is taken care please review

**Comment by niteshsarda** [comment]
I have tested this and **LGTM** for test

**Comment by yvsubhash** [comment]
@ustcweizhou  Please review

File: server/src/com/cloud/storage/StorageManagerImpl.java [hunk 3]
```diff
@@ -2263,27 +2264,43 @@ public void cleanupDownloadUrls(){
...
-
-                s_logger.debug("Removing download url " + volumeOnImageStore.getExtractUrl() + " for volume id " + volumeOnImageStore.getVolumeId());
+                expiredVolumeIds.add(volumeId);
+                s_logger.debug("Removing download url " + volumeOnImageStore.getExtractUrl() + " for volume id " + volumeId);
 
                 // Remove it from image store
                 ImageStoreEntity secStore = (ImageStoreEntity) _dataStoreMgr.getDataStore(volumeOnImageStore.getDataStoreId(), DataStoreRole.Image);
                 secStore.deleteExtractUrl(volumeOnImageStore.getInstallPath(), volumeOnImageStore.getExtractUrl(), Upload.Type.VOLUME);
 
+                _snapshotStoreDao.findByVolume(volumeId, DataStoreRole.Image);
```
Comment (ustcweizhou) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: what's the proposal of line 2285 to line 2287 ?

Comment (yvsubhash) [inline server/src/com/cloud/storage/StorageManagerImpl.java]: @ustcweizhou  That was not needed. So removed that same

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 456
 Hypervisor xenserver
 NetworkType Advanced
 Passed=104
 Failed=1
 Skipped=7

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**
* test_routers_network_ops.py

 * test_03_RVR_Network_check_router_state Failing since 2 runs


**Skipped tests:**
test_01_test_vm_volume_snapshot
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_disk_offerings.py

**Comment by yvsubhash** [comment]
@ustcweizhou does it look good to you now?

**Comment by ustcweizhou** [comment]
code LGTM
 @yvsubhash it would be nice if you can reformat part of the code with java code conventions.

**Comment by yvsubhash** [comment]
@ustcweizhou  coding convention issue is taken care

**Comment by ustcweizhou** [comment]
LGTM now

**Comment by yvsubhash** [comment]
tag:mergeready

**Comment by cloudmonger** [comment]
### ACS CI BVT Run
 **Sumarry:**
 Build Number 535
 Hypervisor xenserver
 NetworkType Advanced
 Passed=112
 Failed=0
 Skipped=7

_Link to logs Folder (search by build_no):_ https://www.dropbox.com/sh/yj3wnzbceo9uef2/AAB6u-Iap-xztdm6jHX9SjPja?dl=0


**Failed tests:**

**Skipped tests:**
test_01_test_vm_volume_snapshot
test_vm_nic_adapter_vmxnet3
test_static_role_account_acls
test_11_ss_nfs_version_on_ssvm
test_nested_virtualization_vmware
test_3d_gpu_support
test_deploy_vgpu_enabled_vm

**Passed test suits:**
test_deploy_vm_with_userdata.py
test_affinity_groups_projects.py
test_portable_publicip.py
test_over_provisioning.py
test_global_settings.py
test_scale_vm.py
test_service_offerings.py
test_routers_iptables_default_policy.py
test_loadbalance.py
test_routers.py
test_reset_vm_on_reboot.py
test_deploy_vms_with_varied_deploymentplanners.py
test_network.py
test_router_dns.py
test_non_contigiousvlan.py
test_login.py
test_deploy_vm_iso.py
test_list_ids_parameter.py
test_public_ip_range.py
test_multipleips_per_nic.py
test_metrics_api.py
test_regions.py
test_affinity_groups.py
test_network_acl.py
test_pvlan.py
test_volumes.py
test_nic.py
test_deploy_vm_root_resize.py
test_resource_detail.py
test_secondary_storage.py
test_vm_life_cycle.py
test_routers_network_ops.py
test_disk_offerings.py

**Comment by yvsubhash** [comment]
@karuturi / @koushik-das  Please merge this

**Pull request merged**